transports/quic/src/muxer.rs

```diff
@@ -0,0 +1,418 @@
+                connection,
+                waker: None,
+                timer: None,
+                substreams: Default::default(),
+                pending_substreams: Default::default(),
+                close_waker: None,
+            }),
+        }
+    }
+
+    pub fn is_handshaking(&self) -> bool {
+        self.inner.lock().connection.is_handshaking()
+    }
+
+    pub fn peer_id(&self) -> Option<PeerId> {
+        let inner = self.inner.lock();
+        let session = inner.connection.crypto_session();
+        C::peer_id(session)
+    }
+
+    pub fn local_addr(&self) -> Multiaddr {
+        let inner = self.inner.lock();
+        let ip = inner
+            .connection
+            .local_ip()
+            .unwrap_or(IpAddr::V4(Ipv4Addr::UNSPECIFIED));
+        let addr = SocketAddr::new(ip, inner.endpoint.port());
+        crate::transport::socketaddr_to_multiaddr(&addr)
+    }
+
+    pub fn remote_addr(&self) -> Multiaddr {
+        let inner = self.inner.lock();
+        let addr = inner.connection.remote_address();
+        crate::transport::socketaddr_to_multiaddr(&addr)
+    }
+
+    pub(crate) fn set_accept_incoming(&self, accept: bool) {
+        let mut inner = self.inner.lock();
+        inner.accept_incoming = accept;
+    }
+}
+
+impl<C: Crypto> StreamMuxer for QuicMuxer<C> {
+    type Substream = StreamId;
+    type OutboundSubstream = ();
+    type Error = QuicMuxerError;
+
+    fn poll_event(
+        &self,
+        cx: &mut Context,
+    ) -> Poll<Result<StreamMuxerEvent<Self::Substream>, Self::Error>> {
+        let mut inner = self.inner.lock();
+        let now = Instant::now();
+
+        while let Poll::Ready(event) = inner.endpoint.poll_channel_events(cx) {
+            inner.connection.handle_event(event);
+        }
+
+        let _max_datagrams = inner.endpoint.max_datagrams();
+        while let Some(transmit) = inner.connection.poll_transmit(now) {
+            inner.endpoint.send_transmit(transmit);
+        }
+
+        loop {
+            if let Some(timer) = inner.timer.as_mut() {
+                match Pin::new(timer).poll(cx) {
+                    Poll::Ready(expired) => {
+                        inner.connection.handle_timeout(expired);
+                        inner.timer = None;
+                    }
+                    Poll::Pending => break,
+                }
+            } else if let Some(when) = inner.connection.poll_timeout() {
+                inner.timer = Some(Timer::at(when));
+            } else {
+                break;
+            }
+        }
+
+        while let Some(event) = inner.connection.poll_endpoint_events() {
+            inner.endpoint.send_endpoint_event(event);
+        }
+
+        while let Some(event) = inner.connection.poll() {
+            match event {
+                Event::HandshakeDataReady => {}
+                Event::Connected => {
+                    // Break here so that the noise upgrade can finish.
+                    return Poll::Pending;
```

Comment:
Can you expand on this? Why does this Future have to pause? Where is the waker of the task registered?

Reply:
this is so the upgrade can complete
https://github.com/libp2p/rust-libp2p/pull/2159/files#diff-b1dcda9367774dd8a742457ff79bc528d030a8cc38d6791efefc78b30f3f2e03R222